kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/data/StringTableTest.kt

```diff
@@ -0,0 +1,213 @@
+         table(headers, file, transform)
+      }
+
+      shouldThrowMessage("Table file must have a header") {
+         val file = resourcesDir.resolve("users-invalid-empty.table")
+         table(headers, file, transform)
+      }
+
+      shouldThrowMessage(
+         """
+         Missing elements from index 2
+         expected:<["id", "username", "fullName"]> but was:<["id", "username"]>
+         """.trimIndent()
+      ) {
+         val file = resourcesDir.resolve("users-invalid-header.table")
+         table(headers, file, transform)
+      }
+   }
+
+   data class UserInfo(val username: String, val fullName: String)
+   val usersTable = table(
+      headers("id", "UserInfo"),
+      row(4, UserInfo("jmfayard", "Jean-Michel Fayard")),
+      row(6, UserInfo("louis", "Louis Caugnault"))
+   )
+
+   context("file.writeTable - success") {
+      val expectedFileContent = """
+id | username | fullName
+4  | jmfayard | Jean-Michel Fayard
+6  | louis    | Louis Caugnault
+      """.trim()
+
+      test("happy path") {
+         val file = tempfile(suffix = ".table")
+         val rows = usersTable.mapRows { (id, userInfo) ->
+            row(id.toString(), userInfo.username, userInfo.fullName)
+         }
+         val fileContent = file.writeTable(headers("id", "username", "fullName"), rows)
+         file.readText() shouldBe expectedFileContent
+         fileContent shouldBe expectedFileContent
+      }
+
+      test("columns should be aligned") {
+         fun row(i: Int): Row3<String, String, String> {
+            val value = "$i".repeat(i)
+            return row(value, value, value)
+         }
+
+         val table = table(
+            headers("a", "b", "c"),
+            row(2),
+            row(4),
+            row(6),
+         )
+         tempfile(suffix = ".table").writeTable(table.headers, table.rows) shouldBe """
+a      | b      | c
+22     | 22     | 22
+4444   | 4444   | 4444
+666666 | 666666 | 666666
+         """.trimIndent()
+      }
+
+      test("| should be escaped") {
+         val table = mapOf("greeting" to "Hello || world").toTable()
+         val file = tempfile(suffix = ".table")
+         file.writeTable(table.headers, table.rows) shouldBe """
+            key      | value
+            greeting | Hello \|\| world
+         """.trimIndent()
+      }
+   }
+
+   context("file.writeTable - validation") {
+      test("Table file must have a .table extension") {
+         shouldThrowMessage(testCase.name.testName) {
+            val fileMissingTableExtension = tempfile()
+            fileMissingTableExtension.writeTable(usersTable.headers, emptyList())
+         }
+      }
+
+      test("Cells con't contain new lines") {
+         val tableWithNewLines =
+            mapOf("1" to "one\n", "two" to "two", "three" to "three\nthree").toTable()
+         shouldThrowMessage(testCase.name.testName) {
+            tempfile(suffix = ".table").writeTable(tableWithNewLines.headers, tableWithNewLines.rows)
+         }
```

Comment:
It's a bit unclear to me what the expected failure message is here

Reply:
so I started with something like

```
context("file.writeTable - validation") {
   test("check extension") {
      shouldThrowMessage("Table file must have a .table extension") {
         val fileMissingTableExtension = tempfile()
         fileMissingTableExtension.writeTable(usersTable.headers, emptyList())
      }
   }
}
```

Then I was like let's give a better name to the test... like the expected error message

```
context("file.writeTable - validation") {
   test("Table file must have a .table extension") {
      shouldThrowMessage("Table file must have a .table extension") {
         val fileMissingTableExtension = tempfile()
         fileMissingTableExtension.writeTable(usersTable.headers, emptyList())
      }
   }
}
```

But then I wondered if there was a way to not duplicate the string?

And sure enough it's easy to do with kotest

```
context("file.writeTable - validation") {
   test("Table file must have a .table extension") {
      shouldThrowMessage(testCase.name.testName) {
         val fileMissingTableExtension = tempfile()
         fileMissingTableExtension.writeTable(usersTable.headers, emptyList())
      }
   }
}
```

I can revert to the first version if you prefer

Reply:
No big deal, lets keep it as is. :)